pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,168 @@
 
 package status
 
-import "sync"
+import (
+	"context"
+	"sync"
+	"time"
 
-type ResourceLock struct {
-	masterLock sync.RWMutex
-	listing    map[interface{}]*sync.Mutex
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"istio.io/pkg/log"
+)
+
+// Task to be performed.
+type Task func(entry *cacheEntry) error
+
+// Worker queue implements an expandable goroutine pool which executes at most one concurrent routine per target
+// resource.  Multiple calls to Push() will not schedule multiple executions per target resource, but will ensure that
+// the single execution uses the latest value.
+type WorkerQueue interface {
+	// Push a task.
+	Push(target Resource, progress Progress)
+	// Run the loop until a signal on the context
+	Run(ctx context.Context)
+	// Delete a task
+	Delete(target lockResource)
```

Comment:
nit: seems like this should take a `Resource` not a `lockResource`, otherwise we are leaking implementation details?